src/gui/mainwindow.cpp

```diff
@@ -1921,6 +1929,44 @@ void MainWindow::pythonDownloadFinished(const Net::DownloadResult &result)
     const Path exePath = result.filePath + u".exe";
     Utils::Fs::renameFile(result.filePath, exePath);
 
+    // verify hash
+    // Python.org only provides MD5 hash but MD5 is already broken and doesn't guarantee file is not tampered.
+    // Therefore, MD5 is only included to prove that the hash is still the same with upstream and we rely on
+    // SHA3-512 for the main check.
+    {
+        QFile file(exePath.data());
+        if (!file.open(QIODevice::ReadOnly))
+        {
+            restoreWidgetsState();
```

Comment:
Wouldn't it be better to use `QScopeGuard`?

Reply:
After adding `verifyPythonInstaller()` it has only 2 instances of it and it doesn't seem worth it.